NDTensors/src/lib/Sectors/src/ordered_sector.jl

```diff
@@ -0,0 +1,42 @@
+struct OrderedSector <: AbstractCategory
+  data::Vector{AbstractCategory}
+end
```

Comment:
I was picturing a design like:
```julia
struct Sector{Data} <: AbstractCategory
  data::Data
end
```
where then it is positional if `Data <: Tuple` and named if `Data <: NamedTuple`. Not sure if that is too naive or has some downsides.

The idea would be that most functions are defined that forward to `data`, and then there are some inner functions defined on either `Tuple` or `NamedTuple` that implement some functionality like fusion.

Reply:
So then, there would be some sort of minimal required interface for what could be used as `data` inside a `Sector`, and precocious users could stick in anything they want as `data` as long as it implements some minimal set of functions.

Reply:
I did this merging of types and got it to work well. Two good things happened:
1. It's no more code than before, and probably slightly less
2. Now both modes of Sector are a subtype of AbstractCategory.

But so far it's been a bit of a failure in terms of really forwarding the behavior to the data (except for some simple things like `length`). One issue preventing this is that iteration over NamedTuples is both kind of strange by default (doesn't give the name info, so I had to work around that slightly) but more fundamentally, I think there has to be some logic about propagating the names forward in the named case for the fusion_rule. So then it prevents such code from working generically for the ordered case where there are no names.

Possibly this could be improved, though, if we introduce a kind of intermediate type or concept like a `Pair{Symbol,<:AbstractCategory}` and also define `fusion_rule` on this and have that handle the name-peeling-off-and-back-on logic at a different level.

Reply:
Right, so basically you were hoping to share more code across the definitions of `fusion_rule(A::NamedSector, B::NamedSector)` and `fusion_rule(o1::OrderedSector, o2::OrderedSector)`? I wasn't expecting there would be much code sharing there because of the very different semantics of `Tuple`/`Vector` and `NamedTuple`.

A slight side note, but one reason I like [Dictionaries.jl](https://github.com/andyferris/Dictionaries.jl) so much is that it defines a dictionary interface which is more closely related to a vector interface in terms of iteration and other operations, so it makes it easier to write code that is generic for both `Vector` and `Dictionary`. It would be nice if there was a `NamedTuple` version of that so it would be easier to right generic code that works for both `Tuple` and `NamedTuple`.

One possibility would be to have a design like:
```julia
fusion_rule(s1::Sector, s2::Sector) = data_fusion_rule(data(s1), data(s2))

data_fusion_rule(s1::Tuple, s2::Tuple) = positional_fusion_rule(s1, s2)
function positional_fusion_rule(s1, s2)
  # Implementation for positional data
end

data_fusion_rule(s1::NamedTuple, s2::NamedTuple) = named_fusion_rule(s1, s2)
function named_fusion_rule(s1, s2)
  # Implementation for named data
end
```
This is kind of a trait-like design/interface.

We could go all-in on a trait-based design with something like this:
```julia
fusion_rule(s1::Sector, s2::Sector) = data_fusion_rule(data(s1), data(s2))

data_fusion_rule(d1, d2) = data_fusion_rule(fusion_style(d1, d2), d1, d2)

abstract type AbstractFusionStyle end
struct PositionalFusion <: AbstractFusionStyle end
struct NamedFusion <: AbstractFusionStyle end

fusion_style(::Tuple) = PositionalFusion()
fusion_style(::NamedTuple) = NamedFusion()

fusion_style(d1, d2) = promote_fusion_style(fusion_style(d1), fusion_style(d2))
promote_fusion_style(s1::AbstractFusionStyle, s2::AbstractFusionStyle) = error("Not implemented")
promote_fusion_style(s1::S, s2::S) where {S <: AbstractFusionStyle} = s1

function data_fusion_rule(::PositionalFusion, d1, d2)
  # Implementation of fusion for positional data
end

function data_fusion_rule(::NamedFusion, d1, d2)
  # Implementation of fusion for named data
end
```
This could allow for some customization of fusion rules when mixing different data types, and then help extend to custom data types like `Dict`/`Dictionary`, etc. It would also allow us to not hard-code definitions of `NamedSector` and `OrderedSector`. It could also be used in other places to provide more fine-grained control of definitions which need to be customized based on the data type, again without using `NamedSector` and `OrderedSector` explicitly, but with the advantage that code can be shared across data structures that do have common interfaces, like `Vector` and `Tuple`.

Reply:
I'm going back and forth in my head about that trait design or a design like this:
```julia
struct PositionalSector{Data} <: AbstractSector
  data::Data
end

struct NamedSector{Data} <: AbstractSector
  data::Data
end
```
and then functionality can just be customized on the `AbstractSector` type, but there could still be flexibility about the `Data` type, where as long as the `data` defined certain functionality, it should "just work", and additionally common functionality like `length` could be defined on `AbstractSector`. However I think it may still be helpful to have some kind of trait system to customize functionality on the `data` since not all `data` will "just work" as a backend for `PositionalSector` or `NamedSector` anyway, so it's not clear to me that design has a big advantage over the trait design I proposed above.

Another design would be:
```julia
struct TupleSector{Data<:Tuple} <: AbstractSector
  data::Data
end

struct NamedTupleSector{Data<:NamedTuple} <: AbstractSector
  data::Data
end

struct VectorSector{Data<:AbstractVector} <: AbstractSector
  data::Data
end

struct DictSector{Data<:AbstractDict} <: AbstractSector
  data::Data
end
```
so then you get even more fine-grained control, but the issue with that for me is that some sector types might share implementations for fusion (like `TupleSector` and `VectorSector`), so then you are left either with making definition on unions of `AbstractSector` to share code or you are back to a trait-like system anyway, and the different `AbstractSector` subtypes aren't so useful since they are really just reflecting the type of the `data` which can be accessed through a trait.